### ISSUE TITLE
[Image_picker]: Image gallery opens twice if we choose select photos permission.

### ISSUE THREAD

**Issue by Trteich** [flutter/flutter#90373]
Hi. Issue https://github.com/flutter/flutter/issues/78108 still present in Image Picker

**Flutter version:** 2.5.1
**image_picker version:** 0.8.4+2
**IOS version:** 14.7.1

**Comment by darshankawar** [comment]
@Trteich 
Can you verify the same on latest master and check if issue still persists or not ?

```
flutter channel master
flutter upgrade

```

**Comment by stuartmorgan** [comment]
@darshankawar Plugins support the latest version of `stable`

**Comment by darshankawar** [comment]
Thanks for the info @stuartmorgan.
I verified this behavior on latest stable and master on iOS and see the same result. ie, After giving `select photos` permission and then trying to select an image from gallery, it opens twice as shown below:


https://user-images.githubusercontent.com/67046386/133999711-1a614641-bcfc-452e-a36d-ae1ce1ab1778.mov


Whereas, giving `allow all access` permission, selecting an image is displayed on main screen as expected:


https://user-images.githubusercontent.com/67046386/133999878-2c7a87fe-03d0-4c2d-ba5b-f0f17b08fbed.mov

Used plugin's official [example](https://pub.dev/packages/image_picker/example) to verify.



<details>
<summary> stable, master flutter doctor -v </summary>

```
[✓] Flutter (Channel stable, 2.5.1, on Mac OS X 10.15.4 19E2269 darwin-x64,
    locale en-GB)
    • Flutter version 2.5.1 at /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffb2ecea52 (2 days ago), 2021-09-17 15:26:33 -0400
    • Engine revision b3af521a05
    • Dart version 2.14.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (4 available)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 93.0.4577.82


• No issues found!

[✓] Flutter (Channel master, 2.6.0-6.0.pre.147, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-GB)
    • Flutter version 2.6.0-6.0.pre.147 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 83dfb2237a (2 days ago), 2021-09-17 23:48:04 -0400
    • Engine revision 31792e0340
    • Dart version 2.15.0 (build 2.15.0-120.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/dhs/Library/Android/sdk
    • Java binary at: /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.    

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/dhs/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/202.7486908/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)        

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[[✓] Connected device (4 available)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 93.0.4577.82


• No issues found!




```

</details>

Labeling it as regression, as it was [fixed](https://github.com/flutter/flutter/issues/78108#issuecomment-842296309) but seem to have surfaced again.

**Comment by balvinderz** [comment]
the first sheet that comes  is for selecting  the photos that can be accessed by the app  and the second sheet that comes is for picking the image @darshankawar.

**Comment by darshankawar** [comment]
@balvinderz Thanks. So is it an expected behavior considering [this](https://github.com/flutter/flutter/issues/78108#issuecomment-842293588) comment as well ?

**Comment by balvinderz** [comment]
I think the expected behaviour should be - the second sheet should only show only those photos that were selected in the first sheet.

**Comment by darshankawar** [comment]
That doesn't seem to be happening at the moment. After selecting first image, then second time, same set of images are being shown. Below screen shows up after selecting first image:

<img width="324" alt="Screenshot 2021-09-21 at 11 55 49 AM" src="https://user-images.githubusercontent.com/67046386/134122012-d3d475a7-c858-40c9-949b-6d870a3939b9.png">

**Comment by balvinderz** [comment]
Yes . I think the expected behaviour should be only those limited photos should be shown.

**Comment by Moluram** [comment]
Hello!
The is happening because shared library is taken before the authorisation level is checked: 
![CodeWithIssue](https://user-images.githubusercontent.com/15612222/135758644-a99f442e-6e0e-4595-b4b0-d215a4e5970b.png)

I have investigated the issue and it seems like it's very hard to achieve.
An Apple Frameworks Engineer answered 11 months ago in [this thread](https://developer.apple.com/forums/thread/663098) that **"Currently it's not possible"**.

If anyone can help with an issue I will appreciate it a lot :)

**Comment by Shashwat2303** [comment]
Hello

**Comment by offline-first** [comment]
Same issue on iOS (14.81) and Plugin version 0.8.4+4.
Any plan to fix it?

**Comment by Pitazzo** [comment]
I confirm this issue is still present

**Comment by om-ha** [comment]
> Hello! The is happening because shared library is taken before the authorisation level is checked: ![CodeWithIssue](https://user-images.githubusercontent.com/15612222/135758644-a99f442e-6e0e-4595-b4b0-d215a4e5970b.png)
> 
> I have investigated the issue and it seems like it's very hard to achieve. An Apple Frameworks Engineer answered 11 months ago in [this thread](https://developer.apple.com/forums/thread/663098) that **"Currently it's not possible"**.
> 
> If anyone could help out with an issue I would appreciate it a lot :)


This issue is still there.

I am already checking permission with [`permission_handler`](https://pub.dev/packages/permission_handler) pub. But regardless whether I do check permission or let `image_picker` checks for permission, we have this problem:

After selecting the image, you get the permission dialog.

I believe root cause is what @Moluram [mentioned here](https://github.com/flutter/flutter/issues/90373#issuecomment-932964519), permission check is done after picking the image.

Note: this duplicate permission prompt only appears one time during app life time, and will not appear until app is killed and re-opened.

**Comment by jmagman** [comment]
cc @hellohuanlin this looks similar to the work you did in camera to explicitly request permissions https://github.com/flutter/flutter/issues/97199

**Comment by hellohuanlin** [comment]
@jmagman yep it seems related. I can take a look.

**Comment by moiye-dev** [comment]
I'm also experiencing this same issue with the image gallery opening twice the very first time on 
iOS version: 15.5. 
Flutter version: 2.5.2
image_picker version: ^0.8.5+3

**Comment by hellohuanlin** [comment]
I did some initial research, looks like it's not related to permission issue similar to the camera plugin. 

Instead this seems to be an expected behavior (first window is for permission's "selected photos" step, and the second window is the actual photo picker, as others have also pointed out. 

I will close it as "won't fix".

**Comment by moiye-dev** [comment]
The problem is that the second time the image gallery is displayed all images are still displayed and allowed access to  and not only the limited set of photos to which the user allowed access to in the first popup.

**Comment by hellohuanlin** [comment]
@moiye-dev [this](https://github.com/flutter/flutter/issues/90373#issuecomment-932964519)

**Comment by jmagman** [comment]
I think the problem is that it's asking for permissions at all with `PHPickerViewController`, which isn't running in the app process.

> It is out-of-process, so photo library access authorization is not needed. 

So this should be updated to just open the `PHPickerViewController` without first checking the `PHAuthorizationStatus`?  https://github.com/flutter/plugins/pull/5915 is changing this logic a bit, we can reassess after that merges.

**Comment by moiye-dev** [comment]
The problem is that the second time the image gallery is displayed all images are still displayed and allowed access to  and not only the limited set of photos to which the user allowed access to in the first popup.

**Comment by hellohuanlin** [comment]
@moiye-dev as @jmagman explained, there is no need to ask for permission to access `PHPickerViewController`, because it's a separate process. So our implementation that request for permission is unnecessary and should be removed.

**Comment by yareyaredesuyo** [comment]
Yes, this library should add permission check and access capabilities and remove permission check from picker methods.

**Comment by migalv** [comment]
Still happening:
iOS version: 15.5
Flutter version: 3.3.5
image_picker version: ^0.8.5+3

**Comment by JustChris** [comment]
WhatsApp does this properly for example.

**Comment by JustChris** [comment]
This is a major issue because apple's app store auditing team will most likely reject those apps, because this is a security issue. It would be much appreciated if you re-prioritize this.

**Comment by jmagman** [comment]
> This is a major issue because apple's app store auditing team will most likely reject those apps, because this is a security issue

There's nothing being accessed that shouldn't be since it's the picker is not running in the app process. It's a (big) annoyance and usability issue but not security issue as far as I know.  I haven't heard of any apps being rejected because of this, but if your app is please let is know!

**Comment by vashworth** [comment]
If you don't need the photo's metadata, you can set `requestFullMetadata` to false, which will remove the prompt for Limited/Full access to Photos.

For example, 
```
_picker.getImageFromSource(
  source: source,
  options: ImagePickerOptions(
    requestFullMetadata: false,
  ),
);
```

Note: Currently, to get an animated gif, you need the photo's metadata and therefore would need `requestFullMetadata` to be true (https://github.com/flutter/flutter/issues/96263).

Explanation: 
As others have stated, this plugin uses `PHPickerViewController`, which does not required permissions to access photos. Permission is only need to access the `PHAsset` (which is basically just metadata about the image). 

In Apple's [2020 WWDC](https://developer.apple.com/videos/play/wwdc2020/10652]) they stated:
> [PHPicker] has privacy built in by default. The app doesn't need direct access to the photos library to show the picker, and the picker won't prompt for access on behalf of your app. It provides the user-selected photos and videos only, which is a clear user intent of sharing the assets with your app. You might wonder, "How do we achieve that?" Well, PHPicker runs out of process of the app even though it looks like PHPicker runs inside the app. It is really running in a separate process from the host app rendered on top of it. But the app can't access the picker directly and not even take screenshots of the picker content. Only what the user actually selects is passed back to the host app.

> The new option [Limited Access] will only allow your app to access a portion of PHAssets in a users photo library. We believe it can enable users to better control their privacy. If a user places your app under Limited Photos Library, there's a couple things you need to keep in mind. First, the picker will still show the entire Photos Library, and all photos and videos can be selected by the user. Second, no matter what user selects in the picker, PHAssets you can access will not change. If you want to request access to additional PHAssets there is a separate API for you to do that.

As I stated above, `PHAsset` is only needed if you need the image's metadata and is determined by `requestFullMetadata`. If `requestFullMetadata` is true, which it currently is by default, this plugin will use the given permissions to fetch the `PHAsset` and return the metadata in the data of the image. If it's unable to get the `PHAsset` due to missing permissions, it will only return the image without the metadata. If `requestFullMetadata` is false, it will not attempt to get the `PHAsset` and only return the image.

Other Notes:
If we were to build a UI on top of it that filters the selection of images to only those selected by the Limited permission, we would lose out on the built-in search and zoom capabilities in `PHPickerViewController`.

Perhaps, `requestFullMetadata` should be false by default? However, we'd have to see if we can get an animated gif without it (https://github.com/flutter/flutter/issues/96263).

**Comment by BenjiFarquhar** [comment]
> If you don't need the photo's metadata, you can set `requestFullMetadata` to false

This fixed the issue for me, thanks.

**Comment by gowthamanvvdn** [comment]
Hi @BenjiFarquhar can you please share working code

**Comment by alperefesahin** [comment]
Any update on this? `requestFullMetadata: false` does not work for me (iPhone 14 pro max- select limited photos)

**Comment by befirst** [comment]
just to link those two: https://github.com/flutter/flutter/issues/65995

**Comment by tobiasht** [comment]
As I understand it, we don't need to ask permission to access images, since all images are visible in the picker, but only the selected one is shared with the app. Then why are we asking the user to select images, isn't the solution to this problem, to show the picker directly, without the "Select Photos..." prompt?

**Comment by idrats** [comment]
Any updates? This bug is really critical

**Comment by Pascal-Orthopy** [comment]
I would also be interested to know if there are already updates here or if the topic is still being worked on at all?

**Comment by parasrastogi** [comment]
I'm facing the same issue. Any workaround or update?

**Comment by kellysjung** [comment]
Hello! I would also like to voice that I am facing the same issue and would like to know if there are updates around this. Thank you!

**Comment by wal33d006** [comment]
This is surprising actually. The amount of time this issue is opened for. The `requestFullMetadata` thing also not works. Do we have any update on this?

**Comment by inamhusain** [comment]
Is there any update on this, or is there any solution to remove the limited access option from that prompt?

**Comment by JohannPie** [comment]
Any news?

**Comment by klncEmre** [comment]
no update yet?

**Comment by Pragt** [comment]
Is there any update on this issue? Does anyone resolve this issue? Appstore has rejected my app due to this one.

**Comment by dev4jam** [comment]
Any ETA on a fix for this issue?

**Comment by rtokun** [comment]
So any eta? it has been already 1 year...

**Comment by jmagman** [comment]
The most current info is still in https://github.com/flutter/flutter/issues/90373#issuecomment-1409660120

**Comment by rtokun** [comment]
@jmagman The requestFullMetada false is not helping with showing the previously selected photos on the limited selection permission.

**Comment by klncEmre** [comment]
You may try to write this part native in Swift guys, i made it on Swift without asking for any full access. Default provided access is private so you don't need to ask for it.

**Comment by buivanday96** [comment]
any update on this?

**Comment by AlfaizTenup** [comment]
any update on this?
@Moluram 
@balvinderz
